### PR TITLE
⚠️ MachinePool API Types and Scaffolding

### PR DESCRIPTION
**Pull request by juan-lee**
**What this PR does / why we need it**:
Adds `MachinePool` types and scaffolding defined in the [MachinePool API proposal](https://github.com/kubernetes-sigs/cluster-api/blob/master/docs/proposals/20190919-machinepool-api.md).

### CODE REVIEW

**Comment by juan-lee** [comment]
```
main.go:62:1: cyclomatic complexity 32 of func `main` is high (> 30) (gocyclo)
func main() {
```
@ncdc @detiber @vincepri Since it's generated code contributing to gocyclo, should we `// nolint:gocyclo` main?

**Comment by ncdc** [comment]
What if we reworked most of the `if` checks from

```go
if err = (&clusterv1alpha2.Cluster{}).SetupWebhookWithManager(mgr); err != nil {
			setupLog.Error(err, "unable to create webhook", "webhook", "Cluster")
			os.Exit(1)
		}
```

to

```go
err = doSomething()
logAndExitIfError(err, "unable to create webhook", "webhook", "Cluster")

func logAndExitIfError(err error, log logr.Logger, msg string, keysAndValues ...interface{}) {
  if err == nil {
    return
  }

  log.Error(msg, keysAndValues...)
  os.Exit(1)
}
```
?

**Comment by juan-lee** [comment]
> What if we reworked most of the `if` checks from
> 
> ```go
> if err = (&clusterv1alpha2.Cluster{}).SetupWebhookWithManager(mgr); err != nil {
> 			setupLog.Error(err, "unable to create webhook", "webhook", "Cluster")
> 			os.Exit(1)
> 		}
> ```
> 
> to
> 
> ```go
> err = doSomething()
> logAndExitIfError(err, "unable to create webhook", "webhook", "Cluster")
> 
> func logAndExitIfError(err error, log logr.Logger, msg string, keysAndValues ...interface{}) {
>   if err == nil {
>     return
>   }
> 
>   log.Error(msg, keysAndValues...)
>   os.Exit(1)
> }
> ```
> 
> ?

@ncdc This seems reasonable to me if we're okay with post `kubebuilder create api` massaging.

**Comment by ncdc** [comment]
> if we're okay with post kubebuilder create api massaging.

We already have to do it for the webhooks because of the check to see if the webhook server is disabled.

Let's let the others chime in w/their thoughts too.

File: api/v1alpha3/machinepool_phase_types.go [hunk 1]
```diff
@@ -0,0 +1,59 @@
+/*
```
Comment (ncdc) [inline api/v1alpha3/machinepool_phase_types.go]: Let's fold this into machinepools_types.go please

File: api/v1alpha3/machinepool_types.go [hunk 2]
```diff
@@ -0,0 +1,167 @@
...
+package v1alpha3
+
+import (
+	corev1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	capierrors "sigs.k8s.io/cluster-api/errors"
+)
+
+const (
+	// MachinePoolFinalizer is set on PrepareForCreate callback.
```
Comment (ncdc) [inline api/v1alpha3/machinepool_types.go]: PrepareForCreate?

File: api/v1alpha3/machinepool_types.go [hunk 3]
```diff
@@ -0,0 +1,167 @@
...
+	// the machine pool to have 100% available capacity. They may either
+	// be machine instances that are running but not yet available or machine instances
+	// that still have not been created.
+	// +optional
+	UnavailableReplicas int32 `json:"unavailableReplicas,omitempty"`
+
+	// FailureReason indicates that there is a problem reconciling the state, and
+	// will be set to a token value suitable for programmatic interpretation.
+	// +optional
+	FailureReason *capierrors.MachinePoolStatusError `json:"errorReason,omitempty"`
```
Comment (ncdc) [inline api/v1alpha3/machinepool_types.go]:
```suggestion
	FailureReason *capierrors.MachinePoolStatusError `json:"failureReason,omitempty"`
```

File: api/v1alpha3/machinepool_types.go [hunk 4]
```diff
@@ -0,0 +1,167 @@
...
+
+	// FailureMessage indicates that there is a problem reconciling the state,
+	// and will be set to a descriptive error message.
+	// +optional
+	FailureMessage *string `json:"errorMessage,omitempty"`
```
Comment (ncdc) [inline api/v1alpha3/machinepool_types.go]:
```suggestion
	FailureMessage *string `json:"failureMessage,omitempty"`
```

File: api/v1alpha3/machinepool_types.go (hunk 2 above)
Comment (juan-lee) [reply api/v1alpha3/machinepool_types.go]: copy/paste error

File: main.go [hunk 5]
```diff
@@ -59,6 +59,7 @@ func init() {
 	// +kubebuilder:scaffold:scheme
 }
 
+// nolint:gocyclo
```
Comment (juan-lee) [inline main.go]: @ncdc temporarily adding this to let tests proceed. Will wait for others to comment on the approach before moving forward.

**Comment by detiber** [comment]
I'm ok with the proposed refactoring. I'm also not against handling that as a followup issue/PR either.

File: api/v1alpha3/machinepool_webhook.go [hunk 6]
```diff
@@ -0,0 +1,58 @@
...
+
+// +kubebuilder:webhook:verbs=create;update,path=/validate-cluster-x-k8s-io-v1alpha3-machinepool,mutating=false,failurePolicy=fail,groups=cluster.x-k8s.io,resources=machinedeployments,versions=v1alpha3,name=validation.machinepool.cluster.x-k8s.io
+// +kubebuilder:webhook:verbs=create;update,path=/mutate-cluster-x-k8s-io-v1alpha3-machinepool,mutating=true,failurePolicy=fail,groups=cluster.x-k8s.io,resources=machinedeployments,versions=v1alpha3,name=default.machinepool.cluster.x-k8s.io
+
+var _ webhook.Defaulter = &MachinePool{}
+var _ webhook.Validator = &MachinePool{}
+
+// Default implements webhook.Defaulter so a webhook will be registered for the type
+func (m *MachinePool) Default() {
+	// TODO(juan-lee): Add machine pool implementation.
```
Comment (detiber) [inline api/v1alpha3/machinepool_webhook.go]: Will these be updated with a followup PR?

Comment (juan-lee) [reply api/v1alpha3/machinepool_webhook.go]: indeed they will @detiber

File: controllers/machinepool_controller.go [hunk 7]
```diff
@@ -0,0 +1,68 @@
...
+		WithOptions(options).
+		Build(r)
+	if err != nil {
+		return errors.Wrap(err, "failed setting up with a controller manager")
+	}
+
+	r.controller = c
+	r.recorder = mgr.GetEventRecorderFor("machinepool-controller")
+	r.config = mgr.GetConfig()
+
```
Comment (vincepri) [inline controllers/machinepool_controller.go]: Nit: remove extra space

File: main.go (hunk 5 above)
Comment (vincepri) [reply main.go]: This is probably ok to keep, given that the function is generated

**Comment by vincepri** [comment]
/approve
/assign @ncdc 

for final lgtm

File: main.go (hunk 5 above)
Comment (ncdc) [reply main.go]: It's fine for now. We can fix in a separate PR.

**Comment by juan-lee** [comment]
> I'm ok with the proposed refactoring. I'm also not against handling that as a followup issue/PR either.

I filed #1946 to follow up on fixing the gocyclo suppression. @ncdc @detiber

**Pull request merged**